atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java

```diff
@@ -190,20 +192,30 @@ protected boolean shouldStopRetrying(int numTimesFailed) {
     public <T, C extends PreCommitCondition, E extends Exception> T runTaskWithConditionThrowOnConflict(
             C condition, ConditionAwareTransactionTask<T, C, E> task) throws E, TransactionFailedRetriableException {
         checkOpen();
-        OpenTransaction openTransaction;
+        try (OpenTransactionImpl openTransaction =
+                runTimed(() -> Iterables.getOnlyElement(startTransactions(ImmutableList.of(condition))), "setupTask")) {
+            return openTransaction.execute(txn -> task.execute(txn, condition));
+        }
+    }
+
+    @Override
+    public List<OpenTransactionImpl> startTransactions(List<? extends PreCommitCondition> conditions) {
         try {
-            openTransaction = runTimed(
-                    () -> Iterables.getOnlyElement(startTransactions(ImmutableList.of(condition))), "setupTask");
+            return startTransactionsInternal(conditions);
         } catch (Exception e) {
-            condition.cleanup();
+            Closer closer = Closer.create(); // N.B. using closer to run all cleanup tasks even if one cleanup throws
+            conditions.forEach(condition -> closer.register(condition::cleanup));
+            try {
+                closer.close();
+            } catch (IOException ex) {
+                e.addSuppressed(ex);
```

Comment:
Should we instead use your ThreadSafeCloser and add a method #closeAfterException(Throwable) or something that will do the logging and #addSuppressed?

I'm also honestly not that fussed with #addSuppressed, it feels like noise. And specifically, in the internal skiing project, might lead to giant stacktraces (and just lots of logging?), because we use pretty large batches for #startTransactions?

How about we log the individual failures somehow as separate loglines